store/store.go

```diff
@@ -437,24 +486,26 @@ func (s *Store) doRequest(client *http.Client, reqOptions *requestOptions, user
 
 	wwwAuth := resp.Header.Get("WWW-Authenticate")
 	if resp.StatusCode == 401 && user != nil && strings.Contains(wwwAuth, "needs_refresh=1") {
-		// close previous response, refresh and retry
+		// close previous response, refresh user and retry
 		resp.Body.Close()
-		err = refreshMacaroon(user)
+		err = s.refreshUser(user)
 		if err != nil {
 			return nil, err
 		}
-		if s.authContext != nil {
-			err = s.authContext.UpdateUser(user)
-			if err != nil {
-				return nil, err
-			}
-		}
-		req, err := s.newRequest(reqOptions, user)
+		// retry
+		return s.doRequest(client, reqOptions, user)
+
+	} else if resp.StatusCode == 401 && strings.Contains(wwwAuth, "refresh_device_session=1") {
```

Comment:
No, right now they can't be returned both at the same time because the store returns the needs refresh response once the first auth fails (device first). It will require some work to verify device and user auth, and return both needs-refresh at the same time. Although it could be a good idea to handle that here now, you think?